### PR TITLE
Add property-no-unknown rule

### PR DESCRIPTION
**Pull request by vio**
- Rel: https://github.com/stylelint/stylelint/issues/950
- Close: https://github.com/betit/stylelint-known-property/issues/6

### CODE REVIEW

File: src/rules/property-no-unknown/__tests__/index.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  }, {
+    code: ".foo { fill: black; }",
+  }, {
+    code: ".foo { -webkit-align-self: center; }",
+  }, {
+    code: ".foo { align-self: center; }",
+  }, {
+    code: ".foo { --bg-color: white; }",
+    description: "ignore standard CSS variables",
+  } ],
```
Comment (evilebottnawi) [inline src/rules/property-no-unknown/__tests__/index.js]: How about hack property `*width: 100px`?

File: src/rules/property-no-unknown/index.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, {
+      actual: enabled,
+      possible: isBoolean,
+    }, {
+      actual: options,
+      possible: {
+        ignoreProperties: [isString],
+      },
+      optional: true,
```
Comment (evilebottnawi) [inline src/rules/property-no-unknown/index.js]: Maybe add `ignoreVendorProperty: true|false` option?

File: src/rules/property-no-unknown/README.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
+# property-no-unknown
+
+Disallow unknown property.
```
Comment (jeddy3) [inline src/rules/property-no-unknown/README.md]: properties

File: src/rules/property-no-unknown/README.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+```css
+.foo { heigth: 100%; }
+/**    ↑
+ *  These properties */
```
Comment (jeddy3) [inline src/rules/property-no-unknown/README.md]: Can we have "These properties" one space over to the left please?

File: src/rules/property-no-unknown/README.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+```
+
+This rule considers properties defined in the [CSS Specifications and browser specific properties](https://github.com/betit/known-css-properties#source) to be known.
+
```
Comment (jeddy3) [inline src/rules/property-no-unknown/README.md]: Can we get a paragraph below this that reads:

"This rule ignores variables (`$sass`, `@less`, `--custom-property`)."

File: src/rules/property-no-unknown/README.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+}
+```
+
+```css
+.foo {
+  align-self: center;
+}
+```
+
+```css
```
Comment (jeddy3) [inline src/rules/property-no-unknown/README.md]: Can you remove this and the next for examples as they can covered with the addition of the "This rule ignores variables (`$sass`, `@less`, `--custom-property`)." sentence.

File: src/rules/property-no-unknown/__tests__/index.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+  reject: [ {
+    code: ".foo { colr: blue; }",
+    message: messages.rejected("colr"),
+    line: 1,
+    column: 8,
+  }, {
+    code: ".foo { -moz-align-self: center; }",
```
Comment (jeddy3) [inline src/rules/property-no-unknown/__tests__/index.js]: Looking through the [tests checklist](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#checklist), can we also get:

1. A test which is rejected on a different line and column to two tests above e.g. `.foo {\n  colr: blue;\n}`
2. A passing and rejected case insensitivity tests. e.g. `.foo { COLOR: blue; }` and `.foo { COLR: blue; }` 

Other than that, it looks like you're succinctly covered off the applicable parts of the checklist.

File: src/rules/property-no-unknown/index.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    })
+
+    if (!validOptions) { return }
+    if (!enabled) { return }
```
Comment (jeddy3) [inline src/rules/property-no-unknown/index.js]: I don't think this or the following conditionals are required.

File: src/rules/property-no-unknown/index.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+  if (isString(ignore)) {
+    return prop === ignore
+  }
+
+  return false
+}
+
+const isPropertyValid = (prop) => {
+  return properties.indexOf(prop) > -1
```
Comment (jeddy3) [inline src/rules/property-no-unknown/index.js]: Idiomatically we tend to write this as `!== -1`.

File: src/rules/property-no-unknown/index.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+const validate = (result, options) => {
+  return function (node) {
+    const prop = node.prop
```
Comment (jeddy3) [inline src/rules/property-no-unknown/index.js]: Idiomatically, we normally include the core logic inside the main code block. Can you move this into the `walkDecls` function all.

Comment (jeddy3) [inline src/rules/property-no-unknown/index.js]: Idiomatically, we use destructuring to access properties e.g. `const { prop } = node`

File: src/rules/property-no-unknown/__tests__/index.js (hunk 1 above)
Comment (jeddy3) [inline src/rules/property-no-unknown/__tests__/index.js]: Yep, can we get a test for this please.

File: src/rules/property-no-unknown/index.js (hunk 2 above)
Comment (jeddy3) [inline src/rules/property-no-unknown/index.js]: I believe `known-css-properties` is a definitive list of properties, including vendor prefixes. And so a `ignoreVendorProperty` option might be redundant.

**Comment by jeddy3** [comment]
@vio It's fantastic to see this PR! Thanks for nailing the list of known properties :)

I've made some requests above, mainly around tests and how we idiomatically structure the stylelint code. It's looking great though!

There's three more places [where the rule needs to be "wired-in"](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#wire-up-the-rule):

- CHANGELOG
- Example config
- List of rules

**Comment by vio** [comment]
Thanks @jeddy3 / @evilebottnawi! Will have a look and will push the required changes ;)

**Comment by vio** [comment]
@jeddy3 I addressed the issues you raised above, let me know if you notice something else.

**Comment by m-allanson** [comment]
Thanks @vio :) I'll let @jeddy3 check that everything is looking good. Just noting that the failing tests here will be fixed when #1638 is merged in.

File: src/rules/property-no-unknown/README.md [hunk 11]
```diff
@@ -1,15 +1,17 @@
...
 ```css
 .foo { heigth: 100%; }
 /**    ↑
- *  These properties */
+ * These properties */
 ```
 
 This rule considers properties defined in the [CSS Specifications and browser specific properties](https://github.com/betit/known-css-properties#source) to be known.
 
+This rule ignores variables ($sass, @less, --custom-property).
```
Comment (jeddy3) [inline src/rules/property-no-unknown/README.md]: Can we get `$sass` etc back-ticked please.

File: src/rules/property-no-unknown/__tests__/index.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    code: ".foo { fill: black; }",
+  }, {
+    code: ".foo { -webkit-align-self: center; }",
+  }, {
+    code: ".foo { align-self: center; }",
+  }, {
+    code: ".foo { --bg-color: white; }",
+    description: "ignore standard CSS variables",
+  }, {
+    code: ".foo { *width: 100px; }",
```
Comment (jeddy3) [inline src/rules/property-no-unknown/__tests__/index.js]: Can we get a companion reject test for this that shows that only the hack part (e.g. the `*`) is ignored, but the property is still tested e.g.:

```js
{
  code: ".foo { *widh: 100px; }",
  message: messages.rejected("widh"),
  line: 1,
  column: 9,
}
```

**Comment by jeddy3** [comment]
> @jeddy3 I addressed the issues you raised above

Nice. Looking really good! I've two more requests: a docs one and an extra test for hacks. 

Thanks for baring with us through these requests. We try to be very thorough when adding new rule as quite often mistakes can only be rectified with breaking-changes to the behaviour of the rule.

As such, we'll need to wait for a 2nd pair of eyes to be casted over the PR once everything is looking ship-shape :)

**Comment by vio** [comment]
Thanks @jeddy3! I pushed the required tweaks.

No worries, my bad for not checking all the requirements.

**Comment by jeddy3** [comment]
Looks great! Lets just wait for those 2nd pair of eyes.

**Comment by ntwb** [comment]
This is fantastic, thanks @vio 👍

**Comment by evilebottnawi** [comment]
LGTM

**Comment by davidtheclark** [comment]
Looks great, thanks @vio!

The rule is pretty straightforward when all the hard work (the listing) is offloaded onto another module!

**Pull request merged**